packages/monitor-v2/src/monitor-og/SnapshotVerification.ts

```diff
@@ -0,0 +1,342 @@
+  try {
+    const response = await fetch(`${url}/${ipfsHash}`);
+    if (!response.ok) {
+      return new Error(`Request on ${response.url} failed with status ${response.status}`);
+    }
+    const data = await response.json();
+
+    // Try to parse the plugins property as JSON.
+    data.data.message.plugins = JSON.parse(data.data.message.plugins);
+    return data;
+  } catch (error) {
+    return new Error(JSON.stringify(error));
+  }
+};
+
+const ipfsMatchGraphql = (ipfsData: IpfsData, graphqlData: GraphqlData): boolean => {
+  const ipfsProposal = ipfsData.data.message;
+  const graphqlProposal = graphqlData.proposals[0];
+
+  // Verify common properties, except for safeSnap plugin.
+  if (
+    ipfsProposal.space !== graphqlProposal.space.id ||
+    ipfsProposal.type !== graphqlProposal.type ||
+    JSON.stringify(ipfsProposal.choices) !== JSON.stringify(graphqlProposal.choices) ||
+    ipfsProposal.start !== graphqlProposal.start ||
+    ipfsProposal.end !== graphqlProposal.end
+  ) {
+    return false;
+  }
+
+  // Verify that the safeSnap plugin exists on both data sources and has the same number of safes.
+  if (
+    ipfsProposal.plugins.safeSnap === undefined ||
+    graphqlProposal.plugins.safeSnap === undefined ||
+    ipfsProposal.plugins.safeSnap.safes.length !== graphqlProposal.plugins.safeSnap.safes.length
+  ) {
+    return false;
+  }
+
+  // Verify that the safes match. We inspect only the properties that are required for verification.
+  for (let i = 0; i < ipfsProposal.plugins.safeSnap.safes.length; i++) {
+    const ipfsSafe = ipfsProposal.plugins.safeSnap.safes[i];
+    const graphqlSafe = graphqlProposal.plugins.safeSnap.safes[i];
+
+    // Verify the network, oSnap address and number of transaction batches.
+    if (
+      ipfsSafe.network !== graphqlSafe.network ||
+      ipfsSafe.umaAddress !== graphqlSafe.umaAddress ||
+      ipfsSafe.txs.length !== graphqlSafe.txs.length
+    ) {
+      return false;
+    }
+
+    // Verify that the transaction batch properties match.
+    for (let j = 0; j < ipfsSafe.txs.length; j++) {
+      const ipfsTransaction = ipfsSafe.txs[j].mainTransaction;
+      const graphqlTransaction = graphqlSafe.txs[j].mainTransaction;
+
+      if (
+        ipfsTransaction.to !== graphqlTransaction.to ||
+        ipfsTransaction.data !== graphqlTransaction.data ||
+        ipfsTransaction.value !== graphqlTransaction.value ||
+        ipfsTransaction.operation !== graphqlTransaction.operation
+      ) {
+        return false;
+      }
+    }
+  }
+
+  return true;
+};
+
+const findSafe = (safe: SafeSnapSafe, chainId: number, ogAddress: string): boolean => {
+  return (
+    safe.network === chainId.toString() && ethersUtils.getAddress(safe.umaAddress) === ethersUtils.getAddress(ogAddress)
+  );
+};
+
+export const verifyProposal = async (
+  transaction: TransactionsProposedEvent,
+  params: MonitoringParams
+): Promise<VerificationResponse> => {
```

Comment:
nit: I think we can make this return type a little less ambiguous.

Right now, VerificationResponse is `{ verified: boolean, error?: string }`.

I think a more specific type would be `{ verified: true } | { verified: false, error: string }`.

This looks more complex, but it allows a caller to check the verified boolean and ts will immediately infer whether error exists or not.

Reply:
totally, will fix